bin/vmware-host-health.rb

```diff
@@ -0,0 +1,185 @@
+
+  option :handlers,
+         description: 'Comma separated list of handlers',
+         long: '--handlers <HANDLER>',
+         proc: proc { |s| s.split(',') },
+         default: []
+
+  option :help,
+         short: '-h',
+         long: '--help',
+         description: 'Show this message',
+         on: :tail,
+         boolean: true,
+         show_options: true,
+         exit: 0
+
+  def send_client_socket(data)
+    if config[:dryrun]
+      puts data.inspect
+    else
+      sock = UDPSocket.new
+      sock.send(data + "\n", 0, '127.0.0.1', 3030)
+    end
+  end
+
+  def send_ok(check_name, msg, source)
+    event = {
+      'name' => check_name,
+      'source' => source,
+      'status' => 0,
+      'output' => "#{self.class.name} OK: #{msg}",
+      'handlers' => config[:handlers]
+    }
+    send_client_socket(event.to_json)
+  end
+
+  def send_warning(check_name, msg, source)
+    event = {
+      'name' => check_name,
+      'source' => source,
+      'status' => 1,
+      'output' => "#{self.class.name} WARNING: #{msg}",
+      'handlers' => config[:handlers],
+      'occurrences' => 2
+    }
+    send_client_socket(event.to_json)
+  end
+
+  def send_critical(check_name, msg, source)
+    event = {
+      'name' => check_name,
+      'source' => source,
+      'status' => 2,
+      'output' => "#{self.class.name} CRITICAL: #{msg}",
+      'handlers' => config[:handlers],
+      'occurrences' => 2
+    }
+    send_client_socket(event.to_json)
+  end
+
+  def send_unknown(check_name, msg)
+    event = {
+      'name' => check_name,
+      'source' => source,
+      'status' => 3,
+      'output' => "#{self.class.name} UNKNOWN: #{msg}",
+      'handlers' => config[:handlers]
+    }
+    send_client_socket(event.to_json)
+  end
+
+  def cache_file(data)
+    file = File.open('/var/log/sensu/vmware-plugin.json', 'w')
+    file.puts data
+    file.close
+  end
+
+  def read_file
+    if File.file?('/var/log/sensu/vmware-plugin.json')
+      file = File.read('/var/log/sensu/vmware-plugin.json')
+      json = JSON.parse(file)
+      return json
+    end
+    {}
+  end
+
+  def run
+    readcache = read_file
+    cachewrite = {}
+    u = URI.parse(config[:host])
+    vim = RbVmomi::VIM.connect host: u.host, user: config[:username], password: config[:password], insecure: config[:verify_ssl]
+    rootfolder = vim.serviceInstance.content.rootFolder
+    dc = rootfolder.childEntity
+    dc.each do |d|
+      d.hostFolder.childEntity.each do |h|
+        h.host.each do |hs|
+          if hs.summary.overallStatus == 'red'
+            hs.triggeredAlarmState.each do |a|
+              msg = "#{hs.name} currently has an alarm #{hs.summary.overallStatus} - #{a.alarm.info.name}"
+              cachewrite["#{hs.name.gsub!(/[^0-9A-Za-z]/, '')}_Alarm"] = hs.name
+              send_critical("#{hs.name.gsub!(/[^0-9A-Za-z]/, '')}_Alarm", msg, hs.name)
+              readcache.delete("#{hs.name.gsub!(/[^0-9A-Za-z]/, '')}_Alarm")
```

Comment:
hmm what happens if you have multiple instances of the check running at once? If someone sets an interval that is lower than it takes to query and process the data from esxi then this  I think if we want to write to the filesystem we need some kind of locking mechanism such as https://ruby-doc.org/core-2.4.1/File.html#method-i-flock which should be supported on all non windows systems I believe.

Reply:
Im not sure I se the need to run multiple of these ate once. Mainly because this is meant as a single check that returns multiple objects similar to a proxy check. Thats the reason I write each item to the socket. So you will ultimately see a device for each datacenter, then see a device for each host in the datacenter. So running multiple of these will only produce the same objects with no increased data. I also dont recommend running this over a 60second interval as that seemed the most efficient for me. Runnign this against our cluster which housed 2 datacenters and 40ish hosts took about 20seconds.